keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,362 @@
+- External Server Access: IPv4-to-IPv4 and IPv6-to-IPv6 access from pods to external servers
+- Ingress Access: Access from IPv4 and/or IPv6 clients to Kubernetes services. Depending on the ingress controller used, connections may cross IP families (IPv4 clients to IPv6 services, or vice versa).
+- Service Discovery: Support for both A and AAAA records for kube-DNS and coreDNS
+- Functionality tested with the Bridge CNI plugin and Host-Local IPAM plugin as references
+- Maintain backwards-compatible support for IPv4-only and IPv6-only clusters
+
+### Non-Goals
+
+- Cross-family connectivity: IPv4-to-IPv6 and IPv6-to-IPv4 connectivity is considered outside of the scope of this proposal, except in the case of ingress access (depending on ingress controller used).
+- CNI network plugins other than the Bridge and Host-Local IPAM plugins should support Kubernetes dual stack, but the development and testing of dual stack support for these other plugins is considered outside of the scope of this proposal.
+- Code changes should support any number IPv4 addresses and any number of IPv6 addresses per pod (up to some reasonable limit). However, testing will be done only up to the following limits:
+  - Pod addresses: 1 IPv4 address and 2 IPv6 addresses per pod maximum
+  - Service addresses: 1 IPv4 and 1 IPv6 service addresses per service maximum
+- For simplicity, for the first release of dual-stack support, the [Kubernetes Endpoints API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpoints-v1-core) will not be modified. Each IPv4/IPv6 address that is assigned to a backend pod will be treated and displayed as a separate endpoint for the corresponding service. Service load balancing (e.g. using kube-proxy) will be performed on a per-address basis, rather than being done on a per-dual-stack-endpoint basis.
+
+## Proposal
+
+In order to support dual-stack in Kubernetes clusters, Kubernetes needs to have awareness of and support dual-stack addresses for pods, nodes and services. To make this happen, the following changes will be required:
+
+- Kubernetes needs to be made aware of multiple IPs per pod (tested with up to one IPv4 and up to two IPv6 addresses).
+- Link Local Addresses (LLAs) on a pod will remain implicit (Kubernetes will not display nor track these addresses).
+- Kubernetes needs to be configurable for up to two service CIDRs.
+- Backend pods for a service can be dual stack. For the first release of dual-stack support, each IPv4/IPv6 address of a backend pod will be treated as a separate Kubernetes endpoint.
+- Kube-proxy needs to support IPv4 and IPv6 services in parallel (e.g. drive iptables and ip6tables in parallel).
+- Health/liveness/readiness probes should work for dual-stack pods. For the first release of dual-stack support, no additional configuration "knobs" will be added for probe definitions, and a probe is deemed successful if either an IPv4 or IPv6 response is received
```

Comment:
That''s not what I'm saying, IIUC your response. I'm thinking that the goal is to see if the pod is working irrespective of the protocol used to communicate with the pod (of which you agree). As such, instead of having knobs to give the user the ability (and complexity) to select what protocol(s) to use, could we just pick a protocol. If only one protocol available, it is the same as today. If both available, pick one, with my preference being IPv6, as that is the end goal (IPv6 only) IMHO.

Later in this doc(L287-290) it talks about having config for both V4 and V6 health checks. I'm thinking we could just keep the original config with single address specified.

Reply:
Here's a corner case to consider. I'm not sure how common this would be, but it may help clarify the tradeoffs we're talking about: Let's say there's a legacy IPv4-only application (one of the reasons we're adding dual stack) that's being run in an otherwise dual-stack cluster. One example is the  standard nginx dockerhub image. If you load this image in an (otherwise) dual-stack cluster, its pod would have both IPv4 and IPv6 addresses assigned, but the application, by default, will only bind to (listen on) IPv4 (bind address of 0.0.0.0, or all v4 IPs). If someone were to try to configure a health probe for this, it would fail if health probes only test for health on IPv6. On the other hand, if we design it so that we're trying health probes on both IPv4 and IPv6 (declaring health if either V4 OR V6 response is received), then the probes would work correctly with out-of-the-box nginx. If this scenario (and others like it) are not a big concern, then limiting probes to V6 only support for dual stack might be a worthwhile tradeoff.

As for kube-proxy's healthz bind address config (L287-290), I don't want to conflate this kube-proxy startup config with the per-probe-definition config. Sure, these are both related to health probes (for kube-proxy, anyway), but the big difference is that the per-probe-definition config knobs are tricky to do in a portable and non-controversial manner, whereas adding v4 and v6 bind address startup configs are fairly trivial, *if* we need them (the *if* being predicated on the tradeoff described above).